concordia/language_model/local_models.py

```diff
@@ -0,0 +1,307 @@
+        seed: int | None = None,
+        system_message = TEXT_SYSTEM_MESSAGE,
+    ) -> str:
+        
+        message = prompt
+        prompt = PROMPT_TEMPLATE.substitute(system_message=system_message, message=message)
+
+        self.logger.debug(f"Sending prompt to LLM: {prompt}")
+        for retry in range(_MAX_SAMPLE_TEXT_ATTEMPTS): 
+            try:
+                response = self._client(
+                    prompt,
+                    stop=terminators,
+                    temperature=temperature,
+                )
+            except ValueError as e:
+                self.logger.error(f"Error while calling LLM with input: {prompt}. Attempt {retry+1} failed.")
+                if retry == _MAX_SAMPLE_TEXT_ATTEMPTS - 1: 
+                    self.logger.error(f"Max retries exceeded. Raising exception.")
+                    raise language_model.InvalidResponseError(prompt)
+            else:
+                self.logger.debug(f"Succeeded after {retry+1} attempts.")
+                self.logger.debug(f"Response from LLM: {response}")
+                break
+
+        if self._measurements is not None:
+            self._measurements.publish_datum(
+                self._channel,
+                {'raw_text_length': len(response)},
+            )
+        return response
+
+    @staticmethod
+    def extract_choices(text):
+        match = re.search(r'\(?(\w)\)', text)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    @override
+    def sample_choice(
+        self,
+        prompt: str,
+        responses: Sequence[str],
+        *,
+        seed: int | None = None,
+    ) -> tuple[int, str, dict[str, float]]:
+        max_characters = len(max(responses, key=len))
+
+        attempts = 1
+        for _ in range(_MAX_MULTIPLE_CHOICE_ATTEMPTS):
+            sample = self.sample_text(
+                prompt,
+                max_characters=max_characters,
+                temperature=0.0,
+                seed=seed,
+                system_message=MULTIPLE_CHOICE_SYSTEM_MESSAGE,
+            )
+            answer = self.extract_choices(sample)
+            try:
+                idx = responses.index(answer)
+            except ValueError:
+                attempts += 1
+                continue
+            else:
+                if self._measurements is not None:
+                    self._measurements.publish_datum(
+                        self._channel, {'choices_calls': attempts}
+                    )
+                debug = {}
+                return idx, responses[idx], debug
+
+        self.logger.error(f"Multiple choice failed after {_MAX_MULTIPLE_CHOICE_ATTEMPTS} attempts.\nLLM Input: {prompt}\nLLM Output: {sample}\nExtracted Answer: {answer}")
+
+        raise language_model.InvalidResponseError(
+             f'Too many multiple choice attempts.\nLLM Input: {prompt}\nLLM Output: {sample}'
+        )
+
+class HuggingFaceLanguageModel(language_model.LanguageModel):
```

Comment:
There is also already a huggingface llm model using AutoModelForCausalLM at HEAD. It's this one:
concordia/language_model/pytorch_gemma_model.py It would probably be best to generalize that model so we don't end up with multiple ways of accomplishing the same thing.

Reply:
I appears that there is an add and then a delete of this file. This file does not seem to exist on my local repo